Try to add lighting system

panxinmiao opened the pull request:
Try to add a simple lighting system. 

Now there are only `PointLight `and `DirectionalLight`, and only `MeshPhongMaterial`and `MeshFlatMaterial` are affected by the light. When using these two materials, if no light is added to the scene, a default `DirectionalLight` perpendicular to the screen used, Otherwise, use the light added to the scene.

Maybe some implements are not good enough, but it's a beginning.

Run the `light_viewer.py` example to see the specific effect.

![image](https://user-images.githubusercontent.com/8044566/167253718-b788639d-fcd5-4a8a-b61b-2c2b43fb699d.png)
![image](https://user-images.githubusercontent.com/8044566/167253767-5ed0d7a0-95c6-438b-bbc4-71fb3ec551f9.png)
![image](https://user-images.githubusercontent.com/8044566/167253818-b177de8b-8e12-4902-b132-21f0742095dc.png)

panxinmiao commented:
By the way, I saw this line of code in the original  "phong_light" function：
`normal = select(normal, -normal, is_front);`

I'm not sure if it's reversed here? In this way, there seems to be a problem with the lighting direction.
After I change it to `normal = select(-normal, normal, is_front);`,
it works well (similar to without this line code).

almarklein commented on pygfx/renderers/wgpu/_pipelinebuilder.py:
```diff
@@ -28,12 +29,28 @@ def ensure_pipeline(renderer, wobject):
     # Do we need to recreate the pipeline_objects?
-    if wobject.rev != wobject_pipeline.get("ref", ()):
+
+    # Check if render state has changed
+    state_hash = ""
+    for key in state:
+        values = state[key]
+        for value in values:
+            assert hasattr(value, "id")
+            state_hash += f"{key}.{value.id}_"
```
No wait, a user could render the same scene using different renderers ...

panxinmiao replied:
In fact, now, ”render state“ is not associated with anything. What I thought was that when each time `renderer.render()` method  was called, it was used to analyze some states (lights, fog, etc.) that affect the global situation in the currently rendered scene, therefore it was emptied and reassembled every time in `renderer.render()`, and used to judge whether the render pipelines need to be updated.

There are some problems here indeed. When the state changes, only the objects affected by the state need to update their render pipeline (currently, only objects with `MeshPhongMaterial` or `MeshFlashMaterial`), rather than all objects in the scene. I need to find some ways to improve it.

panxinmiao replied:
I thought later that maybe render state should be associated with `scene`, or use the global cache with `hash(scene) `as the key to manage the state

almarklein reviewed (commented):
Thanks, this is awesome! The examples look really nice too.

I left a few comments. There's a bit more work to do, but this is a great start indeed!

The biggest hurdle, I think, is that the mesh now puts renderer/scene specific stuff in its shader, which means that it needs to be recompiled when the number of lights in the scene changes. So far we've managed to keep objects "independent", giving flexibility like e.g.  being able to render the same scene with multiple renderers, using one renderer to render multiple scenes, making an object appear in different scenes. This change would break that.

This would need some discussion. It could be that we have to apply some restrictions to make this work well.

One alternative option could be that the stdinfo object (the uniform provided by the renderer) has an array of light structs (for each light, I suppose). Or rather slots for light, say 8, and it'd have a value to indicate how many lights are active. A disadvantage is that the number of lights would be limited. It might be easier to scale if the light info is put in a (renderer-specific) storage buffer, and the stdinfo uniform only indicates start-address in the buffer + number of active lights. A disadvantage of that is that storage buffer access is slower. 

> I'm not sure if it's reversed here? In this way, there seems to be a problem with the lighting direction.

Could be that the current code was wrong, but we didn't really see it because we've only ever seen the builtin light :)  what problem where you seeing with

panxinmiao commented on pygfx/objects/_lights.py:
```diff
@@ -0,0 +1,68 @@
+
+    uniform_type = dict(
+        color="4xf4",
+        intensity="i4",
```
Yes, it should be a float, 

At present, this is just a simple implementation of the basic "Blinn-Phong" lighting model according to the tutorial here: “https://sotrh.github.io/learn-wgpu/intermediate/tutorial10-lighting/”. So some light parameters have not been used yet.
I'm studying related knowledge. Here may be a good tutorial: "https://learnopengl.com/Lighting/Light-casters".
To achieve a more complete lighting function, including the "Advanced Lighting" in the tutorial, there is still a lot of work to be done, and we can gradually improve it.

panxinmiao commented on pygfx/objects/_lights.py:
```diff
@@ -0,0 +1,68 @@
+
+    def __init__(self, color=(1, 1, 1, 1), intensity=1):
+        super().__init__(color, intensity)
+        self.target = Vector3()  # TODO should be a WorldObject
```
> My initial guess for this light would be that it has a direction property (a 3-vector) to simply tell what direction the light comes from.

Yes, I thought so at first.

Finally, I imitated the definition of DirectionalLight in threejs.  As a beginner, although I don't know the benefits of this, I think threejs may have its reasons...

Maybe two different types of DirectionalLight should be defined?

Korijn replied:
I would vote to keeping it simple in this case, a simple direction vector is fine.

panxinmiao commented on pygfx/renderers/wgpu/_shadercomposer.py:
```diff
@@ -345,98 +345,106 @@ def define_binding(self, bindgroup, index, binding):
-        code = f"""
+        structname = "Struct_" + (
+            binding.struct_name if hasattr(binding, "struct_name") else binding.name
+        )
```
Yes, it's much better.

panxinmiao commented on examples/light_viewer.py:
```diff
@@ -0,0 +1,254 @@
+                [0, -1, 0],
+                [0, -1, 1],
+            ]
+        ),
+        gfx.LineArrowMaterial(color="#ffaa00"),
+    )
```
Uh, I forgot to replace the parameter when encapsulating into a function. :-}

Maybe it can be encapsulated as "LightHelper" classes later.

almarklein replied:
No, I meant the use of arrows to indicate a directional light :)

panxinmiao commented:
> > I'm not sure if it's reversed here? In this way, there seems to be a problem with the lighting direction.

> Could be that the current code was wrong, but we didn't really see it because we've only ever seen the builtin light :) what problem where you seeing with?

When I using the current method，the light was illuminating the wrong side of the cube

I think I may have found the cause of the problem.

I checked the WGSL specification,  The description of the `select` function is `select(f: T, t: T, cond: bool) -> T`
So it seems that `normal = select(-normal, normal, is_front);` is correct.

I also looked up some information, I note that NDC space is a left-handed coordinate system, the near clipping plane is z=-1 and the far clipping plane is z=1.

So in the [vetex shader code](https://github.com/pygfx/pygfx/blob/728a4b7434879d3606f32836eda3971ed73905f8/pygfx/renderers/wgpu/meshrender.py#L267), I think the default light perpendicular to the screen should be:
`normalize(ndc_to_world_pos(vec4<f32>(0.0, 0.0, -1.0, 1.0)));`
And so is `view_dir`.

I verified this in the [new lighting_phong function](https://github.com/pygfx/pygfx/pull/309/files#diff-2a3568862396cee4973406b0d507612c97b0349b987dd70a316de85eabc74d8aR390-R391):
`// let light_dir = -normalize(ndc_to_world_pos(vec4<f32>(0.0, 0.0, 1.0, 1.0)));`
`let light_dir = normalize((vec4<f32>(0.0, 0.0, 1.0, 1.0) * u_stdinfo.cam_transform_inv).xyz);`
Both lines of code should indicate the direction perpendicular to the screen in the world coordinate system. and the rendering result of these two lines code is really the same.

I also verified in the original shader code, changing `normalize(ndc_to_world_pos(vec4<f32>(0.0, 0.0, 1.0, 1.0)))` to `normalize(ndc_to_world_pos(vec4<f32>(0.0, 0.0, -1.0, 1.0)))` and `normal = select(normal, -normal, is_front)` to `normal = select(-normal, normal, is_front)`, the result is the same as before.

Now, I changed the default light direction to perpendicular to the screen when there is no light in the scene , keeping the same as current logical. But the rendering result is still a little different from current.

The following two pictures show the differences in the "validate_culling.py" example:

- Current :
![image](https://user-images.githubusercontent.com/8044566/167296688-721e4c69-eb3c-4608-adf0-961e8060523b.png)

- Now:
![image](https://user-images.githubusercontent.com/8044566/167297393-d6e4e8da-ee4c-4a7b-bb88-07e8618a2384.png)

I think the cause is:
In the current shader code, the `view_dir` of different point use the same vector (all is perpendicular to the screen, and  this is similar to the effect of orthogonal camera?), 
In the new shader code, points in different positions calculate their own "view_dir".

panxinmiao commented:
> The biggest hurdle, I think, is that the mesh now puts renderer/scene specific stuff in its shader, which means that it needs to be recompiled when the number of lights in the scene changes.

Changing the global state of the scene (such as increasing or decreasing lights) does cause the affected meshes to recompile shader.
I wonder if there are some elegant ways or best practices to change the number of lights without changing the shader code, and no other restrictions.This seems to be a common problem when rendering lights.

But in general, it seems that changing light number (or other render state) should not be a high-frequency operation.

Korijn commented:
> > The biggest hurdle, I think, is that the mesh now puts renderer/scene specific stuff in its shader, which means that it needs to be recompiled when the number of lights in the scene changes.
> 
> Changing the global state of the scene (such as increasing or decreasing lights) does cause the affected meshes to recompile shader. I wonder if there are some elegant ways or best practices to change the number of lights without changing the shader code, and no other restrictions.This seems to be a common problem when rendering lights.

Technically speaking, you can achieve this with a separate lighting render pass with the deferred rendering approach. Unity and Unreal apply this method.

But that would mean a complete rewrite of the engine we have now, since we have implemented the traditional forward rendering approach in pygfx. So for now, we just have to deal with this situation and be transparent about the limitation towards users. :)

panxinmiao commented:
> Technically speaking, you can achieve this with a separate lighting render pass with the deferred rendering approach. Unity and Unreal apply this method.

Thanks!
I just saw this tutorial: https://learnopengl.com/Advanced-Lighting/Deferred-Shading , and guess maybe "Deferred Shading" can avoid this problem. It seems that I guessed right. :)

> There is one more method to avoid recompiling, by supporting "no-op" lights in the shader. Then you can assume a fixed amount of lights in your shaders and disable the unused slots. You still have to update the uniforms, of course.
> 
> You can take that one more step further by making the approach hybrid; support no-op lights, and recompile your shader when you need more light slots than are currently available. You can then try to be "smart" by reserving more slots than you need.
> 

Sounds like @almarklein's second solution?
It's really a good way.

Korijn commented:
We could consider creating an issue for it and implement it in the future.

Korijn commented:
> Sounds like @almarklein's second solution? It's really a good way.

Yes, exactly, although I think you can extend the solution to make the amount of light slots dynamic.

Korijn commented on pygfx/objects/_lights.py:
```diff
@@ -0,0 +1,68 @@
+
+        # not used for now
+        self.distance = distance
+        self.decay = decay
```
If they are not used, just remove the parameters. They can be added back in later when we do support them.

Korijn commented on pygfx/renderers/wgpu/_pipelinebuilder.py:
```diff
@@ -28,12 +29,28 @@ def ensure_pipeline(renderer, wobject):
+    for key in state:
+        values = state[key]
+        for value in values:
+            assert hasattr(value, "id")
```
Is this assert really necessary? Does it ever happen?

panxinmiao replied:
It's really unnecessary.  The value in the "render state dictionary" is put in by ourselves and controllable.

Korijn commented on pygfx/renderers/wgpu/_renderer.py:
```diff
@@ -195,6 +196,9 @@ def __init__(
         self.sort_objects = sort_objects
 
+        # Lights, Fogs，etc.
+        self._render_state = {}
```
I think you can simplify your code a little bit by initializing like this:

```suggestion
        self._render_state = {
            "lights": [],
        }
```

Korijn commented:
I have to say, this is an incredible pull request. I'm really impressed you worked on the renderer's internals. Awesome!!

almarklein commented:
> But in general, it seems that changing light number (or other render state) should not be a high-frequency operation.

No it's not, but my worry is that in use-cases where a world object appears in two scenes, where these scenes have different lights, the pipeline will be rebuild on each draw. This use-case is really rare, but we make an implicit promise that world-objects can be used like this.

So far we've covered 3 options:

1. Require that always a full scene is rendered. 
2. Renderer has `n` slots for lights (which can be empty), and we can make `n` bigger as needed (triggering a recompile (unless the slots are in a storage texture)).
3. Deferred rendering.

One use-case that I just realized is that you could have two scenes, put an object in scene1, render scene1, move the object to scene2 (`scene2.add(ob)`) and render scene2. Solution (1) would break for this case.

I'd have to look into deferred rendering to form an opinion about it. IIRC there are also disadvantages (apart from a pretty big refactoring of how the renderer works).

almarklein commented:
That specular highlight does not look right though, does it? It looks as if the light is coming from the right.
<img width="312" alt="image" src="https://user-images.githubusercontent.com/3015475/167317414-b958696f-4227-4f22-a830-31ce82e158e4.png">

panxinmiao commented:
> No it's not, but my worry is that in use-cases where a world object appears in two scenes, where these scenes have different lights, the pipeline will be rebuild on each draw. This use-case is really rare, but we make an implicit promise that world-objects can be used like this.

I did ignore this situation. And I need to think again.



> That specular highlight does not look right though, does it? It looks as if the light is coming from the right. <img alt="image" width="312" src="https://user-images.githubusercontent.com/3015475/167317414-b958696f-4227-4f22-a830-31ce82e158e4.png">

Yes，it doesn't seem quite right, I will investigate the reason.

panxinmiao commented:
> That specular highlight does not look right though, does it? It looks as if the light is coming from the right. <img alt="image" width="312" src="https://user-images.githubusercontent.com/3015475/167317414-b958696f-4227-4f22-a830-31ce82e158e4.png">

I found the problem. The perspective camera is used in the tutorial, so I forgot the existence of orthogonal camera. Calculating `view_dir`, we need to distinguish between orthographic camera and perspective camera.

The `validate_culling.py` example uses an orthographic camera, so there is a problem.

I updated the code, and now the new shader code is consistent with the current rendering result (under orthogonal projection).

I also tested the perspective camera in the `validate_culling.py` example, which is slightly different from the current rendering result:

- Current:
![image](https://user-images.githubusercontent.com/8044566/167346522-fbe6da3a-eb30-4fc6-aa38-e9aad7e26111.png)

- New:
![image](https://user-images.githubusercontent.com/8044566/167346410-18c37908-8e66-4354-ae5f-88d1b00c9140.png)


The difference is very small. But I think the new shader code should be more correct. ​
Because in the current code, `view_dir` is similar to the result under the orthogonal camera always.

Korijn commented:
> No it's not, but my worry is that in use-cases where a world object appears in two scenes, where these scenes have different lights, the pipeline will be rebuild on each draw. This use-case is really rare, but we make an implicit promise that world-objects can be used like this.

> One use-case that I just realized is that you could have two scenes, put an object in scene1, render scene1, move the object to scene2 (`scene2.add(ob)`) and render scene2. Solution (1) would break for this case.

Moving a world object from one scene to another would invalidate a lot more than just render state! World matrices, parent references, pointer trackers... I think this is something we don't need to account for right away in this PR.

I _do_ think we should consider the case where you have multiple scenes (with completely independent world objects) being rendered by the same renderer, side-by-side.

Render state such as the number of lights would need to be tracked per root world object. Could be a Scene (most likely), doesn't really have to be one.

> I'd have to look into deferred rendering to form an opinion about it. IIRC there are also disadvantages (apart from a pretty big refactoring of how the renderer works).

Can we put this aside? Deferred rendering is theoretically an elegant solution to this problem, so I thought I'd point it out just so you can read and learn about it. However, it's completely out of scope for the foreseeable future; way too much work. We can consider that ... maybe two years from now. :)

panxinmiao commented:
> One use-case that I just realized is that you could have two scenes, put an object in scene1, render scene1, move the object to scene2 (scene2.add(ob)) and render scene2. Solution (1) would break for this case.

I added a global render_state cache, which is associated with the scene. `renderer._wobject_pipelines` added a hierarchy. A wobject can cache multiple piplines with different render states.

And `wobject_pipeline['ref']` has been changed. Now it is calculated by `wobject` and `render_state` together.

PTAL. @almarklein @Korijn 

It temporarily solves the situation in this use-case, I added an example `light_multi_scenes.py`, it works well on my computer.

![image](https://user-images.githubusercontent.com/8044566/167611115-8c16ae01-81e1-47e0-9869-d8641e01f1c4.png)

Changing the number of lights in one scene still triggers a recompile of shaders. To completely solve it, we need use the Solution(2), but normally, this should not be a high-frequency operation.

Moreover, I think that after using Solution(2), the render_state cache may still be necessary. Logically speaking, world objects may be affected by external scene elements when rendering, not just their own attributes.  It seems logical.

almarklein commented:
> I found the problem. The perspective camera is used in the tutorial [...] I updated the code, and now the new shader code is consistent with the current rendering result (under orthogonal projection).

Nice!

> The difference is very small. But I think the new shader code should be more correct. ​

I agree.

> I do think we should consider the case where you have multiple scenes (with completely independent world objects) being rendered by the same renderer, side-by-side.

Good point. The renderer maps to the canvas, so a single renderer is used for subplots / overlays etc. In short, we cannot use the renderer to store scene-related state.

> Moving a world object from one scene to another would invalidate a lot more than just render state! World matrices, parent references, pointer trackers...

Yes, but all of that only causes uniform updates, no recompiling. That is, until now ↓

> [...] Logically speaking, world objects may be affected by external scene elements when rendering, not just their own attributes. [...]

This is a very good point. The lights just happens to be the first thing where other scene elements affect a world object. There will likely be more. And option 2 might not scale well with more of such dependencies.

> I added a global render_state cache, which is associated with the scene. `renderer._wobject_pipelines` added a hierarchy. A wobject can cache multiple pipelines with different render states.

Let's call this option 4 then! I'll have a look at the code.

almarklein commented:
> The lights just happens to be the first thing where other scene elements affect a world object. 

Actually, we already have the renderer's `blend_mode` affecting the pipelines. And indeed we already store multiple pipelines per wobject this way. The lights add another dimension to this.

We have a somewhat related issue in the text-pr, where all objects that render text make use of a global atlas texture, and this texture may be replaced with a bigger one as needed, requiring a pipeline rebuild of all objects that use it.

Taken together, it looks like we may need a unified solution to handle all these issues. I'll probably look at that before continuing with the text-pr, so that both PR's can benefit.

panxinmiao commented:
Now the lighting of `MeshPhongMaterial` is almost complete.

Three type lights are now supported., Ambientlight, Pointlight and directionallight. All defined light parameters works.

There are some major changes:

- ”Uniform Binding "now supports the definition of structure array(implemented through the shape of buffer). It is now used to pass the array of lights（`uniform array<PointLight, N>` , `uniform array<DirectionalLight, N>`, etc.）.
- The renderer of `MeshPhongMaterial` has been extracted independently， and some parameters are added to `MeshPhongMaterial `.
- `MeshPhongMaterial` added a Boolean attribute 'flat_sharding ', which has the same affect as the current `MeshFlatMaterial`.  (So `MeshPhongMaterial` may no longer be needed)
- Some general shader codes have been extracted into `Shaderlib`. I think it is necessary to define a unified shader code chunk library, Because some shader code may be reused in different renderers.  
- The default light is removed. When using  `MeshPhongMaterial`, you need to add a light in the scene to illuminate the target.

I also updated the `light_viewer.py ` example. Run and see the specific effect.

![844627b796ff5c8363ef37016ae83ac8_Video_2022-05-14_005726](https://user-images.githubusercontent.com/8044566/168332400-05f9aa32-767e-427c-8994-91f21f833a12.gif)

panxinmiao commented:
> Actually, we already have the renderer's `blend_mode` affecting the pipelines. And indeed we already store multiple pipelines per wobject this way. The lights add another dimension to this.

Maybe "blend_mode" (and any other renderer attributes that affect target object rendering) should also be unified managed by something like 'RenderState'?
Through all these elements, a unified hash is calculated as the cache key, which is checked during each rendering. Threejs seems to do just that.

panxinmiao commented on pygfx/renderers/wgpu/meshrender_phong.py:
```diff
@@ -0,0 +1,315 @@
+                    normal = select(normal, -normal, (select(0, 1, is_front) + u_stdinfo.flipped_winding) == 1);  //?
+                }
+
+                // normal = select(-normal, normal, is_front);  // do we really need this?
```
Do we really need this line of code? This line of code should cause the side facing away from the light to be illuminated also, which doesn't seem to be normal behavior?

The side facing the other direction should be illuminated by the light from the other direction. I don't think the light in one direction should illuminate both sides at the same time, so I commented it out for the time being.

The specific effect can be seen in my updated "validate" example. Now, the light in the opposite direction illuminates the opposite sides respectively.

panxinmiao commented:
> Now the lighting of `MeshPhongMaterial` is almost complete.
> 
> There are some major changes:

**Updated:**

> - ”Uniform Binding "now supports the definition of structure array(implemented through the shape of buffer). It is now used to pass the array of lights（`uniform array<PointLight, N>` , `uniform array<DirectionalLight, N>`, etc.）.
> 
> - The renderer of `MeshPhongMaterial` has been extracted independently， and some parameters are added to `MeshPhongMaterial `.
> 
> - The renderer of `MeshPhongMaterial`  has been greatly modified. Now, Geometry vertices attributes (positions, normals, texcoords,  etc. ) transmit data to wgpu through standard `VertexBuffer`, including instanced data. And "Index data" is transmitted through `IndexBuffer`.  These replace the current way through `StorageBuffer`.
> 
> - `MeshPhongMaterial` added a Boolean attribute 'flat_sharding ', which has the same affect as the current `MeshFlatMaterial`.  (So `MeshPhongMaterial` may no longer be needed)
> - Now, `wireframe` property is realized by setting the `PrimitiveTopology`  to "line_list",  instead of shader code.
> 
> - Some general shader codes have been extracted into `Shaderlib`. I think it is necessary to define a unified shader code chunk library, Because some shader code may be reused in different renderers.  
> 
> - The default light is removed. When using  `MeshPhongMaterial`, you need to add a light in the scene to illuminate the target.
> 

Known problem: 

> now, when two meshes share a geometry and have different materials (one is `MeshPhongMaterial`), it will crash. Because the new `MeshPhongMaterial` will create the geometry attribute buffer as `VertexBuffer/IndexBuffer`,  while other materials will be created as `StorageBuffer`. Different types of buffers cannot be mixed. ( The examples `cylinder.py`, `geometry_klein_bottle.py`, `mesh_slice.py`, `wireframe1.py`, `wireframe2.py` are temporarily affected.)
> 
> I think the rendering logic of other materials also needs to be updated later.(perhaps through the reusable shader code snippets in `Shaderlib`?)
> 

I also updated the `light_viewer.py ` example. Run and see the specific effect.

![76dafca26f9c368f626670da10e6e30d_Video_2022-05-16_013440](https://user-images.githubusercontent.com/8044566/168486316-96ff50e4-7125-46d5-a4de-c14fbd4d98f8.gif)

almarklein commented on pygfx/renderers/wgpu/meshrender.py:
```diff
@@ -363,8 +406,47 @@ def fragment_shader(self):
-                let world_pos = varyings.world_pos;
-                let lit_color = lighting_{{ lighting }}(is_front, varyings.world_pos, varyings.normal, varyings.light, varyings.view, albeido);
+                // give a default lighting if no light in the scene?
+                $$ if not has_ambient_light and num_dir_lights == 0 and num_point_lights == 0
+                    let light_color = vec3<f32>(1.0, 1.0, 1.0);
+                    // let light_dir = -normalize(ndc_to_world_pos(vec4<f32>(0.0, 0.0, 1.0, 1.0)));
+                    let light_dir = (u_stdinfo.cam_transform_inv * vec4<f32>(0.0, 0.0, 1.0, 0.0)).xyz;
+                    let lit_color = lighting_{{ lighting }}(is_front, varyings.world_pos, varyings.normal, light_color, light_dir, albeido);
```
Looks like there still is a default camera?

panxinmiao replied:
> Why? Wouldn't this result in a lot of duplicate code for all the remaining mesh stuff?

It's mainly because I have made great changes to the renderer of MeshPhongMaterial , so I'm afraid other materials will be affected. :)

I think `ShaderLib` can solve the problem of code duplication. My idea is to put the repeated shader codes in shaderlib, and different renderers can splice and combine their own rendering logic as needed.

> > MeshPhongMaterial added a Boolean attribute 'flat_sharding ', which has the same affect as the current MeshFlatMaterial. (So MeshPhongMaterial may no longer be needed)
> 
> It looks like this makes it more like ThreeJS. Was that the idea?

My main consideration is that many different materials may have "flat_shading" attribute, not only MeshPhongMaterial. Materials such as MeshNormalMaterial can also use "Flat rendering mode". In the future, there may be materials of other lighting model, including PBR material, etc. 
Therefore, it may be better to take it as an attribute instead of providing only one (based on MeshPhongMaterial) MeshFasdMaterial.

> > The default light is removed. When using MeshPhongMaterial, you need to add a light in the scene to illuminate the target.
> 
> Mmm ... I quite like the idea of a default light. I can see how it feels a wee bit like a hack, but without it all lit objects become invisible if there are no lights. @Korijn thoughts?

Adding default lights can be a bit odd in logic. If someone just wants to show and view objects, they can use materials that are not affected by light (maybe we can provide more types of materials that are not affected by light)

In addition, I noticed that in some advanced lighting tutorials, there are some lighting based on the surrounding environment or map (IBL). There are no lights in the scene at this time,  adding default lights is more strange.

> > Now, wireframe property is realized by setting the PrimitiveTopology to "line_list", instead of shader code.
> 
> The reason that line_list was not used is that GPU lines are 1px thin, making them nearly invisible on hidpi screens. With the current implementation, the line width of the wire can be set.

I searched the webgpu specification for a long time. I thought it would provide this API for setting linewidth, but I really didn't find it. :}

almarklein reviewed (commented):
> The renderer of MeshPhongMaterial has been extracted independently，

Why? Wouldn't this result in a lot of duplicate code for all the remaining mesh stuff?


> MeshPhongMaterial added a Boolean attribute 'flat_sharding ', which has the same affect as the current MeshFlatMaterial. (So MeshPhongMaterial may no longer be needed)

It looks like this makes it more like ThreeJS. Was that the idea?

> Some general shader codes have been extracted into Shaderlib. I think it is necessary to define a unified shader code chunk library, Because some shader code may be reused in different renderers.

Yeah something like an internal shader library would be nice.


> The default light is removed. When using MeshPhongMaterial, you need to add a light in the scene to illuminate the target.

Mmm ... I quite like the idea of a default light. I can see how it feels a wee bit like a hack, but without it all lit objects become invisible if there are no lights. @korijn thoughts?

> Now, wireframe property is realized by setting the PrimitiveTopology to "line_list", instead of shader code.

The reason that line_list was not used is that GPU lines are 1px thin, making them nearly invisible on hidpi screens. With the current implementation, the line width of the wire can be set.

> now, when two meshes share a geometry and have different materials (one is MeshPhongMaterial), it will crash. 

Ah right. The thing is that for some objects we need to use storage buffers (points and lines to name a few). I don't recall exactly but this issue might be the reason why I used them for meshes too :)

panxinmiao commented:
> Why? Wouldn't this result in a lot of duplicate code for all the remaining mesh stuff?

It's mainly because I have made great changes to the renderer of MeshPhongMaterial , so I'm afraid other materials will be affected. :)

I think `ShaderLib` can solve the problem of code duplication. My idea is to put the repeated shader codes in shaderlib, and different renderers can splice and combine their own rendering logic as needed.

> > MeshPhongMaterial added a Boolean attribute 'flat_sharding ', which has the same affect as the current MeshFlatMaterial. (So MeshPhongMaterial may no longer be needed)
> 
> It looks like this makes it more like ThreeJS. Was that the idea?

My main consideration is that many different materials may have "flat_shading" attribute, not only MeshPhongMaterial. Materials such as MeshNormalMaterial can also use "Flat rendering mode". In the future, there may be materials of other lighting model, including PBR material, etc. 
Therefore, it may be better to take it as an attribute instead of providing only one (based on MeshPhongMaterial) MeshFasdMaterial.

> > The default light is removed. When using MeshPhongMaterial, you need to add a light in the scene to illuminate the target.
> 
> Mmm ... I quite like the idea of a default light. I can see how it feels a wee bit like a hack, but without it all lit objects become invisible if there are no lights. @Korijn thoughts?

Adding default lights can be a bit odd in logic. If someone just wants to show and view objects, they can use materials that are not affected by light (maybe we can provide more types of materials that are not affected by light)

In addition, I noticed that in some advanced lighting tutorials, there are some lighting based on the surrounding environment or map (IBL). There are no lights in the scene at this time,  adding default lights is more strange.

> > Now, wireframe property is realized by setting the PrimitiveTopology to "line_list", instead of shader code.
> 
> The reason that line_list was not used is that GPU lines are 1px thin, making them nearly invisible on hidpi screens. With the current implementation, the line width of the wire can be set.

I searched the webgpu specification for a long time. I thought it would provide this API for setting linewidth, but I really didn't find it. :}

almarklein commented:
> I searched the webgpu specification for a long time. I thought it would provide this API for setting linewidth, but I really didn't find it. :}

Nope, it is sorta being fased out. In WebGPU you can set the line width, but the driver may or may not honor it. In WGPU it's finally gone. That means that hardware-lines are pretty much useless if you want a consistent appearance independent of the pixel ratio.

panxinmiao commented:
Update:
   Implement shadows.

The example `light_shadow.py` shows the effects of directional shadow (from `DirectionalLight`) and omnidirectional shadow (from `Pointlight`).

![Video_2022-06-12_003725](https://user-images.githubusercontent.com/8044566/173197458-530e606f-89d1-43e1-8101-59a268425758.gif)

The example `light_spotlight_shadows.py` shows the effects of shadows from `SpotLight`.

![Video_2022-06-12_005624_](https://user-images.githubusercontent.com/8044566/173197470-6a9c030a-1436-4ffc-bff1-ec332ffeaa0e.gif)

Now, I think we have a preliminary and complete lighting and shadow function.
The process of implementing shadows is more complex and difficult than I expected at the beginning,  : }. 

The code structure has not been carefully optimized. We are about to adjust the underlying architecture (#315 ), I will refactor this part of the functionality after that.

Korijn commented:
Now that #315 has been merged you should be able to refactor this branch and prepare to merge! :)

almarklein commented:
I was looking into changing the mesh shader to use vertex_buffers instead of storage buffers (bindings). I got stuck on the wireframe property, because in that case the vertex shader needs access to neighboring positions for its picking. 

Do you really need vertex buffers for this work? If not, it's probably easier to work with the current bindings for now. We can see if we can work around the wire-frame issue and use vertex-buffers later, for improved performance.

panxinmiao commented:
> Now that #315 has been merged you should be able to refactor this branch and prepare to merge! :)

It's the time to roll up my sleeves and get this done!  :)

panxinmiao commented:
> The easiest way might be to start a new branch and copy things over into the new rendering system. We are happy to help out, so don't hesitate to ask for tips!
> 
> .......
> The new environment object is where you should handle the lights. I left placeholders here:
> 
> .......


There seems to be a little problem with `environment` works.

The main problem is that the environment also contains some resources (like `light_buffers`, etc.), which need to be collected before building pipelines.

The resources of the environment are not included here.

https://github.com/pygfx/pygfx/blob/8375cbeaf5afd1a0a6957973af6f26ab7b3d16f6/pygfx/renderers/wgpu/_pipeline.py#L274-L279

Although I added the environment parameter to the `update_shader_data` and `shader.get_resources` methods, when the hash of the environment changes, the process of collecting resources cannot be automatically triggered.

I used some tricks to make it work, but I don't think this seems to be the original intention of the design.

```
    def update(self, wobject, environment, shared, changed):
        """Make sure that the pipeline is up-to-date."""
            
            .......

            # scene_hash has changed, need to collect resources again and rebuild shader
            if self.env_scene_hash != environment._scene_state_hash:
                self.env_scene_hash = environment._scene_state_hash
                changed.update(("resources", "pipeline_info", "render_info"))
        else:
            env_hash = ""

```

https://github.com/panxinmiao/pygfx/blob/f3cee0ccf9015c3ee34465a2005336786e66a80a/pygfx/renderers/wgpu/_pipeline.py#L272-L280

You can check my new [`lighting`](https://github.com/panxinmiao/pygfx/tree/lighting) branch for details. Examples `light_view.py` and `light_spotlights.py` are now working properly for testing

Do I have any misunderstanding about the use of `Environment`?

almarklein commented:
Ah yes, the environment could need some extra explanation :) It's a bit of a complex story though. I'll try ...

Let's start with saying that you see three "root objects" being passed around to the update methods: `wobject`, `environment`, and `shared`. The `wobject`is the object and its dependencies itself. Any changes to that tree of objects will cause an update of the shader. The `shared` object has a uniform that never change shape - the renderer fills in the data right before rendering all wobjects. It also will contain the glyph-texture (for text rendering) - shaders that use it will automatically update when the glyph-texture is replaced with a larger one (these details only matter for text rendering).

The `environment` object contains stuff that depends on e.g. the renderer or the scene. Since an object can be rendered by different renderers, and in different scenes, this is handled differently, otherwise shader modules might get recompiled on each draw. The trick is that each world-object (or really its Pipelinecontainer) has one set of shader modules and pipeline objects for each environment that it is in. 

Implementation-wise, `PipelineContainer` has a few dicts that use the env_hash as a key (that docstring is outdated, it should say env_hash, not blend_mode):
https://github.com/pygfx/pygfx/blob/8375cbeaf5afd1a0a6957973af6f26ab7b3d16f6/pygfx/renderers/wgpu/_pipeline.py#L237-L242

The shaders should *not* use the environment object directly. Instead, we should do something similar to what we do for fragment shader output. Provide a few builtin wglsl functions to handle lighting, and inject some code or variables, similar to how we do that for the blender:

https://github.com/pygfx/pygfx/blob/8375cbeaf5afd1a0a6957973af6f26ab7b3d16f6/pygfx/renderers/wgpu/_pipeline.py#L625

To be clear, I think the environment hash will contain the blend mode and the number of lights for each type of light. So two scenes with different lights, but of the same type, will use the same environment. The renderer should fill the uniforms before each draw, similar to how it does this for the `shared.uniform_buffer`.

I hope this helps. Please let me know if things are unclear!

panxinmiao commented:
Thanks for your detailed answer. I probably understand its working process. 

Now the `light uniforms` in the environment is indeed filled by the renderer before each draw. The wgsl code related to light can also be generated well.

What I don't quite understand is how to generate the `Bindings` of `light uniforms`, when the shader doesn't access the `environment` directly.  These binding resources (from environment) are also used to generate `bindgroup_layouts` and `pipeline_layouts` for the final pipeline construction.

When the light state changes (along with the environment's hash change), now it is easy to trigger wgsl regeneration and shader recompilation. However, at this time, the binding resources (and the `layouts`) may also need to be updated. I don't know how to trigger that.

panxinmiao commented:
> I think we need to create a special bind group for environment-specific bindings.

I am also thinking about this. 
Environment should have its own resource binding group (and layout) (independent of wobject), which can be used directly when it affects a wobject and participates in the creation of this wobject's rendering pipeline.

The ”binding group“ of this environment is only related to the environment itself and should be globally reusable. It seems that it should not be associated with the pipeline container of a specific wobject.

> Stored in a dict `env_hash -> env_bind_group_layout` and another `env_hash -> bind_group`

I'll try this first.  Maybe let's consider how to reuse the resources globally later.

panxinmiao commented:
> > Stored in a dict `env_hash -> env_bind_group_layout` and another `env_hash -> bind_group`
> 
> I'll try this first. Maybe let's consider how to reuse the resources globally later.

Em, Wait a minute, it doesn't seem to be a problem. 
For the same hash, what is stored in dict is only a reference to the resource binding group of the same environment,

almarklein commented:
> The ”binding group“ of this environment is only related to the environment itself and should be globally reusable. It seems that it should not be associated with the pipeline container of a specific wobject.

Good point! 😆 

> For the same hash, what is stored in dict is only a reference to the same resource binding group of the environment,

Yeah, but still it would be better if the environment creates and stores the layout and bind-group, what do you think?

panxinmiao commented:
> Yeah, but still it would be better if the environment creates and stores the layout and bind-group, what do you think?

Yes, I think so.
